device/Microsoft.Azure.Devices.Client/Common/TaskHelpers.cs

```diff
@@ -384,20 +390,25 @@ public static Task WithTimeout(this Task task, TimeSpan timeout, Func<string> er
 
             if (task.IsCompleted || (timeout == Timeout.InfiniteTimeSpan && token == CancellationToken.None))
             {
-                await task;
+                await task.ConfigureAwait(false);
                 return;
             }
 
-            using (var cts = CancellationTokenSource.CreateLinkedTokenSource(token))
+            using (var delayCts = new CancellationTokenSource())
             {
-                if (task == await Task.WhenAny(task, Task.Delay(timeout, cts.Token)))
+                if (task == await Task.WhenAny(task, Task.Delay(timeout, delayCts.Token)).ConfigureAwait(false))
                 {
-                    cts.Cancel();
-                    await task;
+                    delayCts.Cancel();
+                    await task.ConfigureAwait(false);
                     return;
                 }
             }
 
+            using (var operationCts = CancellationTokenSource.CreateLinkedTokenSource(token))
+            {
+                operationCts.Cancel();
+            }
```

Comment:
can you explain why do we need this? what i understand is this will create a cancellation token source and cancel immediately, but it doesn't affect any on original token.